[issue-272] - Pagination Improvement

DeeJayElly opened the pull request:
[issue-272] - Pagination Improvement

Fixes issue #272

a-rukin commented:
<img width="283" alt="image" src="https://user-images.githubusercontent.com/2539345/77408255-7dd1a700-6dc8-11ea-9efd-3389d433612b.png">
on a first page "go first page" should be disabled
plus "go first" "go last" should be a bit different arrows, look at the design

a-rukin commented:
> Works great 🎉 the only issue i found is that the number input is cut-off for larger numbers:
> ![CleanShot 2020-03-24 at 09 32 09](https://user-images.githubusercontent.com/172414/77404953-c67e6580-6db2-11ea-9a30-9cf0bebbf1e7.png)

yeah, it should show full number always
ideally expanding width while typing

DeeJayElly commented:
> Works great 🎉 the only issue i found is that the number input is cut-off for larger numbers:
> ![CleanShot 2020-03-24 at 09 32 09](https://user-images.githubusercontent.com/172414/77404953-c67e6580-6db2-11ea-9a30-9cf0bebbf1e7.png)

Fixed

DeeJayElly commented:
> > Works great 🎉 the only issue i found is that the number input is cut-off for larger numbers:
> > ![CleanShot 2020-03-24 at 09 32 09](https://user-images.githubusercontent.com/172414/77404953-c67e6580-6db2-11ea-9a30-9cf0bebbf1e7.png)
> 
> yeah, it should show full number always
> ideally expanding width while typing

Fixed

DeeJayElly commented:
> <img alt="image" width="283" src="https://user-images.githubusercontent.com/2539345/77408255-7dd1a700-6dc8-11ea-9efd-3389d433612b.png">
> 
> on a first page "go first page" should be disabled
> plus "go first" "go last" should be a bit different arrows, look at the design

Fixed

DominikGuzei commented on source/widgets/browsing/Pagination.tsx:
```diff
@@ -28,7 +47,21 @@ const Pagination = (props: IPaginationProps) => {
-        <span className={styles.pageNumber}>{currentPage}</span>
+        <span className={styles.pageNumber}>
+          <span
+            contentEditable={true}
```
Contenteditable is not recommended with React and this actually has a bug: simply select the current value and replace it with another number (like 20) -> you get "2020" in the editable area (i guess that's because react doesn't know about the edited content, and adds the `currentPage` value.

Please refactor to a normal input component 🙏

DeeJayElly replied:
We can't use input component for this auto width because it cant be styled in the same way like on the design.

With contentEditable i know there are some performance issues, but i tried to catch the onKeyPress event, change the page and stop the rest of event execution with                 e.preventDefault();
e.stopPropagation();

Another thing i am doing is rechecking when the componentDidUpdate and if the property received for currentPage is different then the DOM content of the span which holds the current page, it will just try to force it to the new props and that solved the issue.

In this way we don't have any issue with duplication and current page is always loaded well.

DeeJayElly replied:
Please disregard my last comment. @DominikGuzei

DmitriiGaico commented:
@DeeJayElly Hello! Please see improvements we need for this feature : 
1. Input field should accept Numbers Only. Currently you can enter whatever symbol you want.
![image](https://user-images.githubusercontent.com/45355947/77625809-eea1cc00-6f4c-11ea-9d8c-24b5aa42071f.png)
2.Input Field should be limited.
![image](https://user-images.githubusercontent.com/45355947/77625891-1a24b680-6f4d-11ea-87fd-f6f6bc8ed776.png)
3.While new block is added the Input Field is also Updated.
![Input](https://user-images.githubusercontent.com/45355947/77626156-9a4b1c00-6f4d-11ea-9941-ddbec9fd50b4.gif)
4.Search for the Entered page should be started after Focus is moved Outside of the Input Field and Clicked. Currently Search can be initiated by Hitting Enter.
5.Empty Field "Enter" needs to be restricted.
![EmptyEnter](https://user-images.githubusercontent.com/45355947/77626899-b26f6b00-6f4e-11ea-9547-44dbed27a503.gif)
6.After Hitting enter with Valid Required Page Page shows latest Page and then switched to the Entered one.
![Search](https://user-images.githubusercontent.com/45355947/77627195-2f024980-6f4f-11ea-9235-e0ab4e21cb9b.gif)

DeeJayElly commented:
> @DeeJayElly Hello! Please see improvements we need for this feature :
> 
> 1. Input field should accept Numbers Only. Currently you can enter whatever symbol you want.
>    ![image](https://user-images.githubusercontent.com/45355947/77625809-eea1cc00-6f4c-11ea-9d8c-24b5aa42071f.png)
>    2.Input Field should be limited.
>    ![image](https://user-images.githubusercontent.com/45355947/77625891-1a24b680-6f4d-11ea-87fd-f6f6bc8ed776.png)
>    3.While new block is added the Input Field is also Updated.
>    ![Input](https://user-images.githubusercontent.com/45355947/77626156-9a4b1c00-6f4d-11ea-9941-ddbec9fd50b4.gif)
>    4.Search for the Entered page should be started after Focus is moved Outside of the Input Field and Clicked. Currently Search can be initiated by Hitting Enter.
>    5.Empty Field "Enter" needs to be restricted.
>    ![EmptyEnter](https://user-images.githubusercontent.com/45355947/77626899-b26f6b00-6f4e-11ea-9547-44dbed27a503.gif)
>    6.After Hitting enter with Valid Required Page Page shows latest Page and then switched to the Entered one.
>    ![Search](https://user-images.githubusercontent.com/45355947/77627195-2f024980-6f4f-11ea-9235-e0ab4e21cb9b.gif)

All of the stated issues are now fixed. @DmitriiGaico

a-rukin commented:
<img width="58" alt="image" src="https://user-images.githubusercontent.com/2539345/77696204-08c7c280-6fbe-11ea-9686-68b776e7f89c.png">

One thing I'd improve is to center this text inside input if it's now fixed width

DominikGuzei commented:
Hm … im still seeing the current page input cutoff:
![CleanShot 2020-03-26 at 22 23 37](https://user-images.githubusercontent.com/172414/77698241-94d2ed80-6fb0-11ea-9cbd-008f72c0e96a.png)

DmitriiGaico commented:
@DominikGuzei Here how i see it
![Page_2](https://user-images.githubusercontent.com/45355947/77700079-2dbb3600-6fbc-11ea-96fd-81be36176181.gif)

DeeJayElly commented:
> <img alt="image" width="58" src="https://user-images.githubusercontent.com/2539345/77696204-08c7c280-6fbe-11ea-9686-68b776e7f89c.png">
> 
> One thing I'd improve is to center this text inside input if it's now fixed width

If i do that, there will be movements between the edges of the input field and the actual text. I did that on purpose just so i am sure there is no cut offs inside of input text.

DeeJayElly commented:
> <img alt="image" width="58" src="https://user-images.githubusercontent.com/2539345/77696204-08c7c280-6fbe-11ea-9686-68b776e7f89c.png">
> 
> One thing I'd improve is to center this text inside input if it's now fixed width

OK added

DeeJayElly commented:
> Hm … im still seeing the current page input cutoff:
> ![CleanShot 2020-03-26 at 22 23 37](https://user-images.githubusercontent.com/172414/77698241-94d2ed80-6fb0-11ea-9cbd-008f72c0e96a.png)

Realistically saying how many pages we could get throughout time?

DominikGuzei commented:
@DeeJayElly i would at least anticipate 6 digits, to be safe for a while - the number of pages correlates with the `perPage` param which is constraint to a minimum of 5 per page i think

DominikGuzei commented:
@DeeJayElly @a-rukin @DmitriiGaico i fixed the remaining issue with the paging input.
Please re-test 🙏 

The input should work like this now:
- Only accepts digits
- Submit on "ENTER" key
- Dynamically changes size depending on the number of digits (minimum of 1)
- Restricts number of digits that you can enter based on the total pages

a-rukin commented:
<img width="274" alt="image" src="https://user-images.githubusercontent.com/2539345/78016845-09f94680-7354-11ea-979b-9a94908743f3.png">
@DominikGuzei hmm... I'm a bit confused 
first - aligning 
second - when I click arrow keys, input is not updated 🤔

DmitriiGaico commented:
> <img alt="image" width="274" src="https://user-images.githubusercontent.com/2539345/78016845-09f94680-7354-11ea-979b-9a94908743f3.png">
> 
> @DominikGuzei hmm... I'm a bit confused
> first - aligning (in Safari see screenshot above, in Chrome - ok)
> second - when I click arrow keys, input is not updated 🤔

@DeeJayElly Symbols were changed just fine on this level : 
https://github.com/input-output-hk/cardano-explorer-app/pull/280#issuecomment-604704067

DominikGuzei commented:
@a-rukin please check again, these issues should be fixed 👍
